File: scripts/keep_alive.js
/**
 * Render Backend Keep-Alive Service (Node.js)
 * Automatically queries Render backend services every 10 minutes
 * to keep free instances awake and avoid the 15-minute cold start timeout.
 *
 * Usage:
 *   node scripts/keep_alive.js
 *   node scripts/keep_alive.js https://your-backend.onrender.com
 *   node scripts/keep_alive.js https://app1.onrender.com https://app2.onrender.com
 *
 * Environment variables:
 *   RENDER_BACKEND_URL: Single URL or comma-separated list of URLs
 *   PING_INTERVAL_MINUTES: Defaults to 10
 */

import http from 'node:http';
import https from 'node:https';

const defaultUrls = [
  'https://pet-web-pocz.onrender.com',
  'https://food-website-lozv.onrender.com',
  'https://humlya-salt-product-web.onrender.com',
];

// Read targets from CLI args, env var, or defaults
function getTargetUrls() {
  const cliArgs = process.argv.slice(2).filter((arg) => arg.startsWith('http'));
  if (cliArgs.length > 0) return cliArgs;

  if (process.env.RENDER_BACKEND_URL) {
    return process.env.RENDER_BACKEND_URL.split(',').map((u) => u.trim());
  }

  return defaultUrls;
}

const targetUrls = getTargetUrls();
const intervalMinutes = parseInt(process.env.PING_INTERVAL_MINUTES || '10', 10);
const intervalMs = intervalMinutes * 60 * 1000;

function pingUrl(url) {
  return new Promise((resolve) => {
    const startTime = Date.now();
    const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
    process.stdout.write(`[${timestamp}] Pinging ${url} ... `);

    try {
      const parsedUrl = new URL(url);
      const client = parsedUrl.protocol === 'https:' ? https : http;

      const req = client.get(
        url,
        {
          headers: { 'User-Agent': 'RenderKeepAliveNodeBot/1.0' },
          timeout: 25000,
        },
        (res) => {
          const duration = Date.now() - startTime;
          console.log(`✅ HTTP ${res.statusCode} (${duration}ms)`);
          res.resume(); // Consume stream to free socket
          resolve({ url, status: res.statusCode, duration, ok: true });
        }
      );

      req.on('timeout', () => {
        req.destroy();
        console.log(`⚠️ Timed out after 25s (server may be spinning up)`);
        resolve({ url, status: 'TIMEOUT', duration: 25000, ok: false });
      });

      req.on('error', (err) => {
        console.log(`❌ Error: ${err.message}`);
        resolve({ url, status: 'ERROR', error: err.message, ok: false });
      });
    } catch (e) {
      console.log(`❌ Invalid URL format: ${url}`);
      resolve({ url, status: 'INVALID', error: e.message, ok: false });
    }
  });
}

async function runPingCycle() {
  console.log(`\n--- Ping Delivery Run (${new Date().toLocaleTimeString()}) ---`);
  for (const url of targetUrls) {
    await pingUrl(url);
  }
  console.log(`Next automatic query in ${intervalMinutes} minutes.`);
}

console.log('='.repeat(65));
console.log('🚀 Render Backend Keep-Alive Daemon (Node.js)');
console.log(`🎯 Configured Targets (${targetUrls.length}):`);
targetUrls.forEach((url, i) => console.log(`   ${i + 1}. ${url}`));
console.log(`⏱️  Frequency: Every ${intervalMinutes} minutes (${intervalMs / 1000}s)`);
console.log('='.repeat(65));

// Run immediately on launch
runPingCycle();

// Schedule recurring execution every 10 minutes
setInterval(runPingCycle, intervalMs);

File: scripts/keep_alive.py
"""
Render Backend Keep-Alive Service (Python)
Automatically sends an HTTP GET query to Render backend service(s) every 10 minutes
to prevent the free instance 15-minute idle spin-down.

Usage:
    python scripts/keep_alive.py
    python scripts/keep_alive.py https://your-backend.onrender.com
    python scripts/keep_alive.py https://app1.onrender.com https://app2.onrender.com
"""

import os
import sys
import time
from datetime import datetime
import urllib.request
import urllib.error

DEFAULT_URLS = [
    "https://pet-web-pocz.onrender.com",
    "https://food-website-lozv.onrender.com",
    "https://humlya-salt-product-web.onrender.com",
]

INTERVAL_MINUTES = int(os.getenv("PING_INTERVAL_MINUTES", "10"))
INTERVAL_SECONDS = INTERVAL_MINUTES * 60


def get_target_urls():
    cli_args = [arg for arg in sys.argv[1:] if arg.startswith("http")]
    if cli_args:
        return cli_args

    env_urls = os.getenv("RENDER_BACKEND_URL") or os.getenv("RENDER_BACKEND_URLS")
    if env_urls:
        return [u.strip() for u in env_urls.split(",") if u.strip()]

    return DEFAULT_URLS


def ping_url(url: str):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] Pinging {url} ... ", end="", flush=True)

    req = urllib.request.Request(
        url,
        headers={"User-Agent": "RenderKeepAlivePythonBot/1.0"}
    )

    start_time = time.time()
    try:
        with urllib.request.urlopen(req, timeout=25) as response:
            duration_ms = round((time.time() - start_time) * 1000, 1)
            print(f"✅ Success (HTTP {response.status}) in {duration_ms}ms")
    except urllib.error.HTTPError as e:
        duration_ms = round((time.time() - start_time) * 1000, 1)
        print(f"⚠️ HTTP {e.code} {e.reason} in {duration_ms}ms (Backend is awake and responding)")
    except urllib.error.URLError as e:
        print(f"❌ Connection error: {e.reason}")
    except Exception as e:
        print(f"❌ Unexpected error: {e}")


def main():
    targets = get_target_urls()

    print("=" * 65)
    print("🚀 Render Backend Keep-Alive Daemon (Python)")
    print(f"🎯 Configured Targets ({len(targets)}):")
    for i, t in enumerate(targets, 1):
        print(f"   {i}. {t}")
    print(f"⏱️  Frequency: Every {INTERVAL_MINUTES} minutes ({INTERVAL_SECONDS}s)")
    print("=" * 65)

    while True:
        print(f"\n--- Keep-Alive Query Run ({datetime.now().strftime('%H:%M:%S')}) ---")
        for target in targets:
            ping_url(target)

        print(f"Waiting {INTERVAL_MINUTES} minutes for next cycle...")
        time.sleep(INTERVAL_SECONDS)


if __name__ == "__main__":
    main()
